kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt

```diff
@@ -905,14 +894,31 @@ private fun ObjCExportMapper.canBeInheritedBySameClass(
         return secondClass == firstClass || first !is ConstructorDescriptor && secondClass.isSubclassOf(firstClass)
     }
 
-    return canHaveCommonSubtype(firstClass, secondClass)
+    return firstClass.isSubclassOf(secondClass) || secondClass.isSubclassOf(firstClass)
 }
 
-private fun ObjCExportMapper.canHaveSameSelector(first: FunctionDescriptor, second: FunctionDescriptor): Boolean {
+private fun ObjCExportMapper.canHaveCommonSubtype(first: CallableMemberDescriptor, second: CallableMemberDescriptor, ignoreInterfaceMethodCollisions: Boolean): Boolean {
```

Comment:
It is not clear to me why extracting this method was necessary.
In particular, the refactoring makes it harder to verify that, in the absence of the flag, this PR doesn't alter the behavior.

What was wrong with simply adding a condition here? https://github.com/JetBrains/kotlin/blob/e2640b4afac5b4e6e2eb4aa17165593a764cb715/kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt#L885

Reply:
I extracted the function because the name was misleading, the function was checking if the classes were already in the same hierarchy or if they could be in the same hierarchy. 

I can go back to just plumbing the condition all the way into this method though.

Reply:
> I extracted the function because the name was misleading

Which name?

Reply:
`canHaveCommonSubtype` is misleading since it checks whether the classes are in the same hierarchy. IIUC if two classes are in the same hierarchy then they trivially can have a common sub-type which made reasoning about the code a little difficult since this change only cares about the cases where the types aren't in the same hierarchy

Reply:
> `canHaveCommonSubtype` is misleading since it checks whether the classes are in the same hierarchy

Sorry, I don't understand what you mean.

`canHaveCommonSubtype` checks if a Kotlin class can be declared as a subtype of both input classes (considering that `Foo` is a subtype of `Foo` from the compiler point of view).

Reply:
Or do you mean that adding the flag check into `canHaveCommonSubtype` makes the name misleading?
In that case, there is still an easy way to make it less misleading -- replace the call site, `canHaveCommonSubtype(firstClass, secondClass)`, with two cases and flag check?

Reply:
I think that I'll leave the functions as they were previously and plumb the flag all the way down the call stack. This will have the least impact on the existing functionality/code.